File: src/siminov/connect/service/AsyncServiceWorker.java
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package siminov.connect.service;

import java.util.Iterator;

import siminov.connect.IRequest;
import siminov.connect.IWorker;
import siminov.connect.connection.ConnectionManager;
import siminov.connect.connection.design.IConnectionResponse;
import siminov.connect.exception.ConnectionException;
import siminov.connect.exception.ServiceException;
import siminov.connect.model.ServiceDescriptor;
import siminov.connect.model.ServiceRequestResource;
import siminov.connect.resource.ResourceManager;
import siminov.connect.resource.ResourceUtils;
import siminov.connect.service.design.IService;
import siminov.connect.utils.Utils;
import siminov.core.database.design.IDatabase;
import siminov.core.exception.DatabaseException;
import siminov.core.exception.SiminovCriticalException;
import siminov.core.exception.SiminovException;
import siminov.core.log.Log;
import siminov.core.utils.ClassUtils;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

/**
 * It provides implementation for IWorker.
 * It processes all service ASYNC requests
 */
public class AsyncServiceWorker implements IWorker {

	private static AsyncServiceWorker asyncServiceWorker = null;
	
	private AsyncServiceWorkerThread asyncServiceWorkerThread = null;
	private ResourceManager resourceManager = null;
	
	private AsyncServiceWorkerHelper serviceUtils = null;
	
	private ConnectivityChangeReceiver connectivityChangeReceiver = new ConnectivityChangeReceiver();

	/**
	 * AsyncServiceWorker Constructor
	 */
	private AsyncServiceWorker() {
		
		resourceManager = ResourceManager.getInstance();
		serviceUtils = new AsyncServiceWorkerHelper();
		
		startWorker();
		
		/*
		 * Register Connectivity Change Receiver.
		 */
		siminov.core.resource.ResourceManager ormResourceManager = siminov.core.resource.ResourceManager.getInstance();
		Context applicationContext = ormResourceManager.getApplicationContext();
		
		applicationContext.registerReceiver(connectivityChangeReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
	}
	
	
	/**
	 * It provides AsyncServiceWorker singleton instance
	 * @return AsyncServiceWorker singleton instance
	 */
	public static AsyncServiceWorker getInstance() {
		
		if(asyncServiceWorker == null) {
			asyncServiceWorker = new AsyncServiceWorker();
		}
		
		return asyncServiceWorker;
	}
	
	public void addRequest(final IRequest request) {
		
		IService iService = (IService) request;
		
		siminov.connect.model.ServiceRequest service = serviceUtils.convert(iService);
		boolean contain = false;
		
		try {
			contain = serviceUtils.containService(service);
		} catch(ServiceException se) {
			Log.error(AsyncServiceWorker.class.getName(), "process", "ServiceException caught while checking exsisting service, " + se.getMessage());
			iService.onTerminate(se);
			
			return;
		}
		
		if(contain) {
			return;
		}
			
		
		try {
			service.save();
		} catch(SiminovException se) {
			Log.error(AsyncServiceWorker.class.getName(), "process", "SiminovException caught while saving service, " + se.getMessage());
			iService.onTerminate(new ServiceException(AsyncServiceWorker.class.getName(), "process", se.getMessage()));
			
			return;
		}
		
		
		/*
		 * Service Queued
		 */
		iService.onQueue();
		/*
		 * Service Paused
		 */
		iService.onPause();
		
		
		/*
		 * Notify Async Service Worker Thread.
		 */
		startWorker();
	}

	public void removeRequest(final IRequest service) {
		
	}
	
	public boolean containsRequest(final IRequest request) {
		
		IService service = (IService) request;
		siminov.connect.model.ServiceRequest serviceRequest = serviceUtils.convert(service);
		
		try {
			return serviceUtils.containService(serviceRequest);
		} catch(ServiceException se) {
			Log.error(AsyncServiceWorker.class.getName(), "containsRequest", "ServiceException caught while checking exsisting service, " + se.getMessage());
			service.onTerminate(se);
			
			return false;
		}
	}
	
	public void startWorker() {
		
		if(asyncServiceWorkerThread == null) {
			asyncServiceWorkerThread = new AsyncServiceWorkerThread();
		}
		
		if(!asyncServiceWorkerThread.isAlive()) {
			asyncServiceWorkerThread.start();
		}
	}
	
	public void stopWorker() {
		
		if(asyncServiceWorkerThread == null) {
			return;
		}
		
		try {
			asyncServiceWorkerThread.interrupt();
			asyncServiceWorkerThread = null;
		} catch(Exception e) {
			Log.error(AsyncServiceWorkerThread.class.getName(), "stop", "Exception caught while stopping async service worder thread, " + e.getMessage());
			return;
		}
	}
	
	public boolean isWorkerRunning() {
		
		if(asyncServiceWorkerThread == null) {
			return false;
		}
		
		return asyncServiceWorkerThread.isAlive();
	}
	
	/**
	 * It is the inner class of Async service worker which processes all the requests in the queue
	 */
	private class AsyncServiceWorkerThread extends Thread {
		
		public void run() {
		
			siminov.connect.model.ServiceRequest[] services = null;
			try {
				services = new siminov.connect.model.ServiceRequest().select().execute();
			} catch(SiminovException se) {
				Log.error(AsyncServiceWorkerThread.class.getName(), "run", "SiminovException caught while getting queue services, " + se.getMessage());
				throw new SiminovCriticalException(AsyncServiceWorkerThread.class.getName(), "run", "SiminovException caught while getting queue services, " + se.getMessage());
			}
			
			if(services == null || services.length <= 0) {
				
				AsyncServiceWorker.getInstance().stopWorker();
				return;
			}

			for(int i = 0;i < services.length;i++) {
				siminov.connect.model.ServiceRequest service = services[i];
				IService iService = null;
				
				try {
					iService = serviceUtils.convert(service);
				} catch(ServiceException se) {
					Log.error(AsyncServiceWorkerThread.class.getName(), "run", "ServiceException caught while converting service to iService, " + se.getMessage());
					return;
				}

				
				/*
				 * Check Network Connectivity.	
				 */
				if(!Utils.hasCoverage()) {
					
					try {
						this.wait();
					} catch(Exception e) {
						Log.error(AsyncServiceWorkerThread.class.getName(), "run", "Exception caught while putting async service worker thread into wait state, " + e.getMessage());
						return;
					}
				}
				
				
				/*
				 * Service Resumed
				 */
				iService.onResume();
				handle(iService);
			}
			
			AsyncServiceWorker.getInstance().stopWorker();
		}
		
		private void handle(final IService iService) {
			
			IConnectionResponse connectionResponse = null;
			
			try {
				connectionResponse = ConnectionManager.getInstance().handle(iService);
			} catch(ConnectionException se) {
				Log.error(AsyncServiceWorker.class.getName(), "handle", "SiminovException caught while invoking connection, " + se.getMessage());
				
				iService.onTerminate(new ServiceException(se.getClassName(), se.getMethodName(), se.getMessage()));
				return;
			}
			
			iService.onRequestFinish(connectionResponse);
			
			IDatabase serviceRequest = serviceUtils.convert(iService);
			
			try {
				serviceRequest.delete().execute();
			} catch(DatabaseException de) {
				Log.error(AsyncServiceWorker.class.getName(), "handle", "Database Exception caught while deleting service request from database, " + de.getMessage());
				throw new SiminovCriticalException(AsyncServiceWorker.class.getName(), "handle", "Database Exception caught while deleting service request from database, " + de.getMessage());
			}
		}
	}

	private class ConnectivityChangeReceiver extends BroadcastReceiver {

		public void onReceive(Context context, Intent intent) {
			
			if(intent.getExtras() == null) {
				return;
			}
			
			
			NetworkInfo networkInfo = (NetworkInfo) intent.getExtras().get(ConnectivityManager.EXTRA_NETWORK_INFO);
			if(networkInfo.isConnected()) {

				if(asyncServiceWorkerThread != null && asyncServiceWorkerThread.isAlive()) {
					asyncServiceWorkerThread.interrupt();
				}
			} else {
				
				if(asyncServiceWorkerThread.isAlive()) {
					
					try {
						asyncServiceWorkerThread.wait();
					} catch(Exception e) {
						Log.error(AsyncServiceWorkerThread.class.getName(), "onReceive", "Exception caught while putting async service worker thread into wait state because of no connectivity, " + e.getMessage());	
						return;	
					}
				}
			}
		}
	}

	
	/**
	 * It is the helper class for AsyncServiceWorker
	 * It helps to convert request database instance to IService instance and vis versa
	 */
	private class AsyncServiceWorkerHelper {
		
		/**
		 * Check whether it contains the requested service or not
		 * @param service Service
		 * @return (true/false) TRUE: If service request already exists | FALSE: If service does not exists
		 * @throws ServiceException If there is any exception while checking for request
		 */
		public boolean containService(siminov.connect.model.ServiceRequest service) throws ServiceException {
			
			siminov.connect.model.ServiceRequest[] services = null;
			try {
				services = new siminov.connect.model.ServiceRequest().select().execute();
			} catch(DatabaseException de) {
				Log.error(AsyncServiceWorker.class.getName(), "containService", "DatabaseException caught while getting services from database, " + de.getMessage());
				throw new ServiceException(AsyncServiceWorker.class.getName(), "containService", de.getMessage());
			}
			
			
			if(services == null || services.length <= 0) {
				return false;
			}
			
			for(int i = 0;i < services.length;i++) {
				
				siminov.connect.model.ServiceRequest savedService = services[i];
				if(service.getService().equalsIgnoreCase(savedService.getService())) {
					
					if(service.getRequest().equalsIgnoreCase(savedService.getRequest())) {
						
						boolean contain = true;
						
						Iterator<ServiceRequestResource> serviceRequestResources = service.getServiceRequestResources();
						while(serviceRequestResources.hasNext()) {
							
							ServiceRequestResource serviceResource = serviceRequestResources.next();
							ServiceRequestResource savedRequestResource = savedService.getServiceRequestResource(serviceResource.getName());
							
							if(savedRequestResource == null) {
								contain = false;
								break;
							}
							
							if(!serviceResource.getName().equalsIgnoreCase(savedRequestResource.getName())) {
								contain = false;
								break;
							} else if(!serviceResource.getValue().equalsIgnoreCase(savedRequestResource.getValue())) {
								contain = false;
								break;
							}
						}
						
						if(contain) {
							return true;
						}
					}
				} 
			}
			
			
			return false;
		}
		
		/**
		 * Converts the service request database instance to IService instance
		 * @param service ServiceRequest instance
		 * @return IService instance
		 * @throws ServiceException If any exception occur while converting the instance
		 */
		public IService convert(siminov.connect.model.ServiceRequest service) throws ServiceException {
			
			IService iService = (IService) ClassUtils.createClassInstance(service.getInstanceOf());
			iService.setRequestId(service.getRequestId());
			iService.setService(service.getService());
			iService.setRequest(service.getRequest());
			
			Iterator<ServiceRequestResource> serviceRequestResources = service.getServiceRequestResources();
			while(serviceRequestResources.hasNext()) {
				ServiceRequestResource serviceResource = serviceRequestResources.next();
				iService.addResource(serviceResource.getName(), serviceResource.getValue());
			}

			ServiceDescriptor serviceDescriptor = resourceManager.requiredServiceDescriptorBasedOnName(service.getService());
			iService.setServiceDescriptor(serviceDescriptor);
			
			ResourceUtils.resolve(iService);

			return iService;
		}
		
		/**
		 * It converts IService instance to database service request instance
		 * @param iService IService instance
		 * @return ServiceRequest Instance
		 */
		public siminov.connect.model.ServiceRequest convert(final IService iService) {
			
			siminov.connect.model.ServiceRequest serviceRequest = new siminov.connect.model.ServiceRequest();
			serviceRequest.setRequestId(iService.getRequestId());
			serviceRequest.setService(iService.getService());
			serviceRequest.setRequest(iService.getRequest());
			serviceRequest.setInstanceOf(iService.getClass().getName());
			
			Iterator<String> resources = iService.getResources();
			while(resources.hasNext()) {
				String resourceName = resources.next();
				Object resourceValue = iService.getResource(resourceName);
				
				if(!(resourceValue instanceof String)) {
					continue;
				}
				
				ServiceRequestResource serviceRequestResource = new ServiceRequestResource();
				serviceRequestResource.setServiceRequest(serviceRequest);
				serviceRequestResource.setName(resourceName);
				serviceRequestResource.setValue((String) resourceValue);
				
				serviceRequest.addServiceRequestResource(serviceRequestResource);
			}
			
			return serviceRequest;
		}
	}
} 

File: Windows/universal8.1/Siminov/Connect/Service/AsyncServiceWorker.cs
///
/// [SIMINOV FRAMEWORK - CONNECT]
/// Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.



using Siminov.Connect.Connection;
using Siminov.Connect.Connection.Design;
using Siminov.Connect.Exception;
using Siminov.Connect.Model;
using Siminov.Connect.Resource;
using Siminov.Connect.Service.Design;
using Siminov.Core.Database.Design;
using Siminov.Core.Exception;
using Siminov.Core.Log;
using Siminov.Core.Utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Siminov.Connect.Service
{

    /// <summary>
    /// It provides implementation for IWorker.
    /// It processes all service ASYNC requests
    /// </summary>
    public class AsyncServiceWorker : IWorker
    {
        
	    private static AsyncServiceWorker asyncServiceWorker = null;
	
	    private Task asyncServiceWorkerThread = null;
	    private static ResourceManager resourceManager = null;
	
	    private static AsyncServiceWorkerHelper serviceUtils = null;
	
	    private ConnectivityChangeReceiver connectivityChangeReceiver = new ConnectivityChangeReceiver();

        /// <summary>
        /// AsyncServiceWorker Constructor
        /// </summary>
	    private AsyncServiceWorker() 
        {
		
		    resourceManager = ResourceManager.GetInstance();
		    serviceUtils = new AsyncServiceWorkerHelper();
		
		    StartWorker();
		
		    /*
		     * Register Connectivity Change Receiver.
		     */
		    //Core.Resource.ResourceManager coreResourceManager = Core.Resource.ResourceManager.GetInstance();
		
		    //applicationContext.registerReceiver(connectivityChangeReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
	    }
	
        /// <summary>
        /// It provides AsyncServiceWorker singleton instance
        /// </summary>
        /// <returns>AsyncServiceWorker singleton instance</returns>
	    public static AsyncServiceWorker GetInstance() 
        {
		
		    if(asyncServiceWorker == null) 
            {
			    asyncServiceWorker = new AsyncServiceWorker();
		    }
		
		    return asyncServiceWorker;
	    }
	
	    public void AddRequest(IRequest request) 
        {
		
		    IService iService = (IService) request;
		
		    Connect.Model.ServiceRequest service = serviceUtils.Convert(iService);
		    bool contain = false;
		
		    try 
            {
			    contain = serviceUtils.ContainService(service);
		    } 
            catch(ServiceException se) 
            {
			    Log.Error(typeof(AsyncServiceWorker).Name, "Process", "ServiceException caught while checking exsisting service, " + se.GetMessage());
			    iService.OnTerminate(se);
			
			    return;
		    }
		
		    if(contain) 
            {
			    return;
		    }
			
		
		    try 
            {
			    service.Save();
		    } 
            catch(SiminovException se) 
            {
			    Log.Error(typeof(AsyncServiceWorker).Name, "Process", "SiminovException caught while saving service, " + se.GetMessage());
			    iService.OnTerminate(new ServiceException(typeof(AsyncServiceWorker).Name, "Process", se.GetMessage()));
			
			    return;
		    }
		
		
		    /*
		     * Service Queued
		     */
		    iService.OnQueue();
		    /*
		     * Service Paused
		     */
		    iService.OnPause();
		
		
		    /*
		     * Notify Async Service Worker Thread.
		     */
		    StartWorker();
	    }

	    public void RemoveRequest(IRequest service) 
        {
		
	    }
	
	    public bool ContainsRequest(IRequest request) 
        {
		
		    IService service = (IService) request;
		    Connect.Model.ServiceRequest serviceRequest = serviceUtils.Convert(service);
		
		    try 
            {
			    return serviceUtils.ContainService(serviceRequest);
		    } 
            catch(ServiceException se) 
            {
			    Log.Error(typeof(AsyncServiceWorker).Name, "ContainsRequest", "ServiceException caught while checking exsisting service, " + se.GetMessage());
			    service.OnTerminate(se);
			
			    return false;
		    }
	    }
	
	    public void StartWorker() 
        {

            if (asyncServiceWorkerThread == null || asyncServiceWorkerThread.IsCompleted || asyncServiceWorkerThread.IsCanceled || asyncServiceWorkerThread.IsFaulted) 
            {
			    asyncServiceWorkerThread = new Task(new Action(HandleRequests));
		    }
            else
            {
                return;
            }
		
		    //if(!asyncServiceWorkerThread.Status. != TaskStatus.Running) 
            {
			    asyncServiceWorkerThread.Start();
		    }
	    }
	
	    public void StopWorker() 
        {
		
		    if(asyncServiceWorkerThread == null) 
            {
			    return;
		    }
		
		    try 
            {
			    //asyncServiceWorkerThread.interrupt();
			    asyncServiceWorkerThread = null;
		    } 
            catch(System.Exception e) 
            {
                Log.Error(this.GetType().Name, "Stop", "Exception caught while stopping async service worder thread, " + e.Message);
			    return;
		    }
	    }
	
	    public bool IsWorkerRunning() 
        {
		
		    if(asyncServiceWorkerThread == null) 
            {
			    return false;
		    }
		
		    return asyncServiceWorkerThread.IsCompleted;
	    }
	
        /// <summary>
        /// It is the inner class of Async service worker which processes all the requests in the queue
        /// </summary>
		public void HandleRequests() 
        {
		
			Connect.Model.ServiceRequest[] services = null;
			try 
            {
                services = (Connect.Model.ServiceRequest[]) new Connect.Model.ServiceRequest().Select().Execute();
			} 
            catch(SiminovException se) 
            {
                Log.Error(this.GetType().Name, "HandleRequests", "SiminovException caught while getting queue services, " + se.GetMessage());
                throw new SiminovCriticalException(this.GetType().Name, "HandleRequests", "SiminovException caught while getting queue services, " + se.GetMessage());
			}
			
			if(services == null || services.Length <= 0) 
            {
				
				AsyncServiceWorker.GetInstance().StopWorker();
				return;
			}

			for(int i = 0;i < services.Length;i++) 
            {
				Connect.Model.ServiceRequest service = services[i];
				IService iService = null;
				
				try 
                {
					iService = serviceUtils.Convert(service);
				} 
                catch(ServiceException se) 
                {
                    Log.Error(this.GetType().Name, "HandleRequests", "ServiceException caught while converting service to iService, " + se.GetMessage());
					return;
				}

				
	    		/*
				* Check Network Connectivity.	
				*/
				if(!Utils.Utils.HasCoverage()) 
                {
					
					try 
                    {
						asyncServiceWorkerThread.Wait();
					} 
                    catch(System.Exception e) 
                    {
                        Log.Error(this.GetType().Name, "HandleRequests", "Exception caught while putting async service worker thread into wait state, " + e.Message);
						return;
					}
				}
				
				
				/*
				    * Service Resumed
				    */
				iService.OnResume();
				Handle(iService);
			}
			
			AsyncServiceWorker.GetInstance().StopWorker();
		}
		
		private void Handle(IService iService) 
        {
			
			IConnectionResponse connectionResponse = null;
			
			try 
            {
				connectionResponse = ConnectionManager.GetInstance().Handle(iService);
			} 
            catch(ConnectionException se) 
            {
				Log.Error(typeof(AsyncServiceWorker).Name, "Handle", "SiminovException caught while invoking connection, " + se.GetMessage());
				
				iService.OnTerminate(new ServiceException(se.GetClassName(), se.GetMethodName(), se.GetMessage()));
				return;
			}
			
			iService.OnRequestFinish(connectionResponse);
			
			IDatabase serviceRequest = serviceUtils.Convert(iService);
			
			try 
            {
				serviceRequest.Delete().Execute();
			} 
            catch(DatabaseException de) 
            {
				Log.Error(typeof(AsyncServiceWorker).Name, "Handle", "Database Exception caught while deleting service request from database, " + de.GetMessage());
				throw new SiminovCriticalException(typeof(AsyncServiceWorker).Name, "Handle", "Database Exception caught while deleting service request from database, " + de.GetMessage());
			}
		}

	    private class ConnectivityChangeReceiver //: BroadcastReceiver 
        {

		    /*public void onReceive(Context context, Intent intent) 
            {
			
			    if(intent.getExtras() == null) 
                {
				    return;
			    }
			
			
			    NetworkInfo networkInfo = (NetworkInfo) intent.getExtras().get(ConnectivityManager.EXTRA_NETWORK_INFO);
			    if(networkInfo.isConnected()) 
                {

				    if(asyncServiceWorkerThread != null && asyncServiceWorkerThread.isAlive()) 
                    {
					    asyncServiceWorkerThread.interrupt();
				    }
			    } 
                else 
                {
				
				    if(asyncServiceWorkerThread != null && asyncServiceWorkerThread.isAlive()) 
                    {
					
					    try 
                        {
						    asyncServiceWorkerThread.wait();
					    } 
                        catch(System.Exception e) 
                        {
						    Log.Error(typeof(AsyncServiceWorkerThread).Name, "onReceive", "Exception caught while putting async service worker thread into wait state because of no connectivity, " + e.GetMessage());	
						    return;	
					    }
				    }
			    }
		    }*/
	    }

	
        /// <summary>
        /// It is the helper class for AsyncServiceWorker
        /// It helps to convert request database instance to IService instance and vis versa
        /// </summary>
	    private class AsyncServiceWorkerHelper 
        {
		
		    /**
		     * Check whether it contains the requested service or not
		     * @param service Service
		     * @return (true/false) TRUE: If service request already exists | FALSE: If service does not exists
		     * @throws ServiceException If there is any exception while checking for request
		     */
		    public bool ContainService(Connect.Model.ServiceRequest service) 
            {
			
			    Connect.Model.ServiceRequest[] services = null;
			    try 
                {
                    services = (Connect.Model.ServiceRequest[]) new Connect.Model.ServiceRequest().Select().Execute();
			    } 
                catch(DatabaseException de) 
                {
				    Log.Error(typeof(AsyncServiceWorker).Name, "ContainService", "DatabaseException caught while getting services from database, " + de.GetMessage());
				    throw new ServiceException(typeof(AsyncServiceWorker).Name, "ContainService", de.GetMessage());
			    }
			
			
			    if(services == null || services.Length <= 0) 
                {
				    return false;
			    }
			
			    for(int i = 0;i < services.Length;i++) 
                {
				
				    Connect.Model.ServiceRequest savedService = services[i];
				    if(service.GetService().Equals(savedService.GetService(), StringComparison.OrdinalIgnoreCase)) 
                    {
					
					    if(service.GetRequest().Equals(savedService.GetRequest(), StringComparison.OrdinalIgnoreCase)) 
                        {
						
						    bool contain = true;
						
						    IEnumerator<ServiceRequestResource> serviceRequestResources = service.GetServiceRequestResources();
						    while(serviceRequestResources.MoveNext()) 
                            {
							
							    ServiceRequestResource serviceResource = serviceRequestResources.Current;
							    ServiceRequestResource savedRequestResource = savedService.GetServiceRequestResource(serviceResource.GetName());
							
							    if(savedRequestResource == null) 
                                {
								    contain = false;
								    break;
							    }
							
							    if(!serviceResource.GetName().Equals(savedRequestResource.GetName(), StringComparison.OrdinalIgnoreCase)) 
                                {
								    contain = false;
								    break;
							    } 
                                else if(!serviceResource.GetValue().Equals(savedRequestResource.GetValue(), StringComparison.OrdinalIgnoreCase)) 
                                {
								    contain = false;
								    break;
							    }
						    }
						
						    if(contain) 
                            {
							    return true;
						    }
					    }
				    } 
			    }
			
			
			    return false;
		    }
		
            /// <summary>
            /// Converts the service request database instance to IService instance
            /// </summary>
            /// <param name="service">ServiceRequest instance</param>
            /// <returns>IService instance</returns>
            /// <exception cref="Siminov.Connect.Exception.ServiceException">ServiceException If any exception occur while converting the instance</exception>
		    public IService Convert(Connect.Model.ServiceRequest service) 
            {
			
			    IService iService = (IService) ClassUtils.CreateClassInstance(service.GetInstanceOf());
			    iService.SetRequestId(service.GetRequestId());
			    iService.SetService(service.GetService());
			    iService.SetRequest(service.GetRequest());
			
			    IEnumerator<ServiceRequestResource> serviceRequestResources = service.GetServiceRequestResources();
			    while(serviceRequestResources.MoveNext()) 
                {
				    ServiceRequestResource serviceResource = serviceRequestResources.Current;
				    iService.AddResource(serviceResource.GetName(), serviceResource.GetValue());
			    }

			    ServiceDescriptor serviceDescriptor = resourceManager.RequiredServiceDescriptorBasedOnName(service.GetService());
			    iService.SetServiceDescriptor(serviceDescriptor);
			
			    ResourceUtils.Resolve(iService);

			    return iService;
		    }
		

            /// <summary>
            /// It converts IService instance to database service request instance
            /// </summary>
            /// <param name="iService">IService instance</param>
            /// <returns>ServiceRequest Instance</returns>
		    public Connect.Model.ServiceRequest Convert(IService iService) 
            {
			
			    Connect.Model.ServiceRequest serviceRequest = new Connect.Model.ServiceRequest();
			    serviceRequest.SetRequestId(iService.GetRequestId());
			    serviceRequest.SetService(iService.GetService());
			    serviceRequest.SetRequest(iService.GetRequest());
			    serviceRequest.SetInstanceOf(iService.GetType().Name);
			
			    IEnumerator<String> resources = iService.GetResources();
			    while(resources.MoveNext()) 
                {
				    String resourceName = resources.Current;
				    Object resourceValue = iService.GetResource(resourceName);
				
				    if(!(resourceValue is String)) 
                    {
					    continue;
				    }
				
				    ServiceRequestResource serviceRequestResource = new ServiceRequestResource();
				    serviceRequestResource.SetServiceRequest(serviceRequest);
				    serviceRequestResource.SetName(resourceName);
				    serviceRequestResource.SetValue((String) resourceValue);
				
				    serviceRequest.AddServiceRequestResource(serviceRequestResource);
			    }
			
			    return serviceRequest;
		    }
	    }
    }
}
